### ISSUE TITLE
Feature requests for the google life sciences backend

### ISSUE THREAD

**Issue by johanneskoester** [snakemake/snakemake#457]
The Google Life Science API backend for Snakemake, developed by @vsoch and funded by Google, has been released some time ago. While the cooperation continues, we are happy to consider feature requests regarding Google cloud support by the community.

Please post your feature requests here.

**Comment by aryam7** [comment]
Hello Johannes and Vanessa,
Thank you for adding the Google Life Sciences backend to Snakemake!
Do you offer support for running rules within [preemtible VM instances](https://cloud.google.com/life-sciences/docs/reference/gcloud-examples#using_preemptible_vms), by any chance? Perhaps this could be specified as a resource, like
```
rule a:
    output:
        "test.txt"
    resources:
        preemptible=True
    shell:
        "somecommand ..."
```

**Comment by vsoch** [comment]
Yes, it looks like we can! https://cloud.google.com/life-sciences/docs/reference/rest/v2beta/projects.locations.pipelines/run#VirtualMachine I'll get in a PR to test this week.

**Comment by vsoch** [comment]
Update: please see https://github.com/snakemake/snakemake/pull/549 for the preemptible instance option.

**Comment by kdbinder** [comment]
@johanneskoester @vsoch 

Would be ideal to automatically log stderr into a logging bucket within GCS instead of relying on the end user to add a "log" statement in their Snakefile

**Comment by aryam7** [comment]
I'll echo @kdbinder's request. The `stderr` from `gcloud beta lifesciences operations describe` is often only partial. And if the error happens prior to the execution of the rule, the log output (from using the `log` directive) will be empty. This can make it difficult to debug problems that might be occurring due to the Life Sciences Executor.
But maybe I'm missing something. @vsoch, how did you debug the executor when you were first developing it?

The Google Life Sciences API seems to automatically mount [a `/google` virtual disk with complete log output](https://cloud.google.com/life-sciences/docs/reference/rest/v2beta/projects.locations.pipelines/run#action) whenever it executes a job. Would it be possible to copy the contents of `/google` into the storage bucket specified by the user? For example, the files could go into the `source` directory in a folder named by the life sciences operation ID, like this
```
snakemake-testing-data/source/logs/11698975339184312706
```

This option could be enabled via a command line flag, like `--google-lifesciences-logs`, which takes as an optional argument the list of rules whose log output should be copied. If no rules are specified, it could just copy the logs of all of the rules.

**Comment by aryam7** [comment]
If anybody else needs to access the complete logs, you can try [this script](https://gist.github.com/aryam7/8ab14ef42de9f085ecd47f2409520c15) I wrote. It recreates the failed google life sciences submission and copies the log output to a storage bucket of your choosing (under `/google/logs`). It also opens an ssh server with access to the job as it is running.
(Thanks to @moschetti for helping me write it!)

**Comment by vsoch** [comment]
Thanks @aryam7 that is really useful! I think we should add it proper as a feature, I'm waiting for @johanneskoester to give his thoughts before going ahead with a PR.

**Comment by aryam7** [comment]
## Yet another feature request
### Current behavior
For each rule in the Snakefile, Snakemake's Life Sciences Executor currently instructs the Life Sciences API to
1. Create a VM with disk space and memory specified by the [`disk_mb` and `mem_mb` resources](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#resources)
2. Pull a container with `Snakemake` and `conda`
3. Within the container, execute commands to download
    a. the user's working directory files (ie scripts, conda environment files, etc) via [this script](https://gist.github.com/vsoch/84886ef6469bedeeb9a79a4eb7aec0d1)
    b. the input files for the rule [via `GS`](https://github.com/snakemake/snakemake/blob/master/snakemake/remote/GS.py)

Please correct me if I'm misunderstanding anything in the steps above!
### Requested behavior
I'm proposing to alter step 3b.
It seems like the Life Sciences API can mount storage buckets via FUSE (if you set the `enable_fuse` option to true in the [request's Action](https://cloud.google.com/life-sciences/docs/reference/rpc/google.cloud.lifesciences.v2beta#action)). Instead of downloading the input, could we just access it through a FUSE mount?
### Why request this?
1. In some cases, it can be very hard to predict the amount of space required to store the input and output files on the VM. Users currently must either set `disk_mb` to something really large or attempt a process of trial and error, where the `disk_mb` value is carefully tuned for each rule (and each `sample` wildcard for that rule). Even if the developer of a Snakefile gets this to work with their example data, the users of that Snakefile might have much larger files.
2. Under the current implementation, all required input and output for a rule must be specified explicitly. Otherwise, it won't be accessible to the rule in the container. For example, `.bam.bai` index files will need to be specified in each rule's input/output where they weren't really before. This isn't too hard to do, but with FUSE, the file system should behave more naturally.

To illustrate how much time this might save the average Snakefile developer: I've been plagued by issues that are, for the most part, variants of these two for the past week. Granted, these types of things should become faster to debug with improved logging.

**Comment by vsoch** [comment]
I think I remember discussing this with @johanneskoester and @moschetti and we decided against it (or it could have been a different method toward the same goal?) and I can't remember the details now! So @aryam7 you are suggesting that we would need to still upload these inputs to storage on the first go, but then create a filesystem, add the files to it, and then mount the filesystem to all steps and clean up at the end? This feels like overkill since we can just upload once and then access from storage. I do see your point about the redundancy, and I'm concerned that creating this extra step would require making another instance just to populate files, and then waiting for that.

> Under the current implementation, all required input and output for a rule must be specified explicitly. Otherwise, it won't be accessible to the rule in the container. For example, .bam.bai index files will need to be specified in each rule's input/output, unlike on a regular on-premises file system.

I think it's better practice for the user to explicitly state inputs and outputs for different steps - it's more transparent to know what is going on, so I'm not sure this is a strong rationale to add this complexity. But hmm, I see what you mean... what if instead there was a way to define a reference (and then use it in several spots?) Another issue that I see is that it's straight forward to reference a storage file - it's a global identifier in a way. But then once you add a filesystem and path, it's no longer really that.

> To illustrate how much time this might save the average Snakefile developer: I've been plagued by issues that are, for the most part, variants of these two for the past week.

Haha, I hope you are ok! I think this is good to discuss because it's come up before. Could you provide a few examples of these issues, explicitly, maybe that is the best place to start?

**Comment by aryam7** [comment]
Admittedly, this is starting to push the limits of my (just recently acquired) knowledge about anything google cloud, but here goes...
> you are suggesting that we would need to still upload these inputs to storage on the first go, but then create a filesystem, add the files to it, and then mount the filesystem to all steps and clean up at the end?

I'm actually not sure what you mean by _filesystem_ here. Why would you have to create one, add files to it, and then clean up at the end? I thought that every time Snakemake submits a life sciences request, it would just have to do the following in addition to what it usually does:
1. Create an action with `enable_fuse` and `run_in_background` set to true using a container with `gcsfuse` pre-installed
2. In the commands for the action, [mount the storage bucket](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/docs/mounting.md#mounting)
3. Add the mount path to every other action that would usually be submitted in the request

> I think it's better practice for the user to explicitly state inputs and outputs for different steps - it's more transparent to know what is going on, so I'm not sure this is a strong rationale to add this complexity.

That's a good point. I suppose I just thought that if you can easily make the filesystem act more naturally, why not do it? ...
So after receiving your response, I decided to read a bit more about GCS FUSE and figure out if there are any disadvantages to using it like this. Turns out there are [quite a few](https://cloud.google.com/storage/docs/gcs-fuse#notes). Most notably, random reads are slow and concurrent writes aren't supported. I'm not a filesystems expert, though, so maybe somebody else can speak to the significance of these better than I can.
> Could you provide a few examples of these issues, explicitly, maybe that is the best place to start?

The primary reason I submitted this feature request is that it can be hard for me to know how much disk space my rules require, especially when their inputs (and outputs) might differ in size depending on the value of the rules' wildcards. So I thought: why not just mount a storage bucket with almost limitless disk space?
But given the filesystem limitations of GCS FUSE, I'm now thinking it might be better if I just build some code into the beginning of my Snakefile to somehow evaluate the file size of the inputs for each of my wildcards and then set the `disk_mb` values using a lambda.

**Comment by moschetti** [comment]
Yeah, there are a number of performance limitations with GCS FUSE. In addition, you'd need to mount it in the same action that needs access as you can't share the mounts between containers since the API doesn't allow you access directly to the host. So the logic for doing the mount or not would need to be incorporated into the Snakemake container.

**Comment by johanneskoester** [comment]
> Thanks @aryam7 that is really useful! I think we should also add it proper as a feature, I'm waiting for @johanneskoester to give his thoughts before going ahead with a PR.

@vsoch Go for it. Maybe put the logs under .snakemake/google-life-sciences-logs instead of a new toplevel folder, analogously to the current .snakemake/logs, where all the terminal output of Snakemakes master process is repeated. And then display a message at end of execution and on every error that shows where the logs are.

**Comment by johanneskoester** [comment]
> Admittedly, this is starting to push the limits of my (just recently acquired) knowledge about anything google cloud, but here goes...
> 
> > you are suggesting that we would need to still upload these inputs to storage on the first go, but then create a filesystem, add the files to it, and then mount the filesystem to all steps and clean up at the end?
> 
> I'm actually not sure what you mean by _filesystem_ here. Why would you have to create one, add files to it, and then clean up at the end? I thought that every time Snakemake submits a life sciences request, it would just have to do the following in addition to what it usually does:
> 
>     1. Create an action with `enable_fuse` and `run_in_background` set to true using a container with `gcsfuse` pre-installed
> 
>     2. In the commands for the action, [mount the storage bucket](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/docs/mounting.md#mounting)
> 
>     3. Add the mount path to every other action that would usually be submitted in the request
> 
> 
> Am I misunderstanding things here? Or are we actually talking about the same thing?
> 
> > I think it's better practice for the user to explicitly state inputs and outputs for different steps - it's more transparent to know what is going on, so I'm not sure this is a strong rationale to add this complexity.
> 
> That's a good point. I suppose I just thought that if you can easily make the filesystem act more naturally, why not do it? ...
> So after receiving your response, I decided to read a bit more about GCS FUSE and figure out if there are any disadvantages to using it like this. Turns out there are [quite a few](https://github.com/GoogleCloudPlatform/gcsfuse#performance). Most notably, random reads are slow and concurrent writes aren't supported. I'm not a filesystems expert, though, so maybe somebody else can speak to the significance of these better than I can.
> 
> > Could you provide a few examples of these issues, explicitly, maybe that is the best place to start?
> 
> The primary reason I submitted this feature request is that it can be hard for me to know how much disk space my rules require, especially when their inputs (and outputs) might differ in size depending on the value of the rules' wildcards. So I thought: why not just mount a storage bucket with almost limitless disk space?
> But given the filesystem limitations of GCS FUSE, I'm now thinking it might be better if I just build some code into the beginning of my Snakefile to somehow evaluate the file size of the inputs for each of my wildcards and then set the `disk_mb` values using a lambda. Thanks for taking the time to help me work through this, though!

I agree with @vsoch's thoughts. Having input and output explicit is good in any case, and the fuse approach has too many downsides.

Regarding your disk_mb problem, you need to add --default-resources to the snakemake invocation. This way, disk_mb will be automatically set to a reasonable default (`max(2*input.size_mb, 1000)`). The same for mem_mb.

**Comment by johanneskoester** [comment]
Btw. @vsoch can you please add a very prominent section about --default-resources to the google life sciences docs?

**Comment by aryam7** [comment]
> Regarding your disk_mb problem, you need to add --default-resources to the snakemake invocation. This way, disk_mb will be automatically set to a reasonable default (max(2*input.size_mb, 1000)). The same for mem_mb.

Ah! I didn't know you could do that. This is exactly what I need. Thank you!

**Comment by vsoch** [comment]
hey @johanneskoester @kdbinder @moschetti I want to clarify a few points before starting work! In the early comments I think we were talking about getting logs from the instance and putting them into Google Storage, but @johanneskoester you mention writing to the local machine in the .snakemake folder. We know that the logs are generated automatically for the [action as a mount](https://cloud.google.com/life-sciences/docs/reference/rpc/google.cloud.lifesciences.v2beta#action) and so what I'd like to discuss is 

 1) how and where we trigger some upload of the logs, and 
 2) where they ultimately go

@johanneskoester the snakemake exec command is currently the last thing we run for the container entrypoint, so should there be a flag added to that to `--save--google-lifesciences-logs`  ? It probably makes sense to add an upload of the log files to the same spot that would upload results on complete - can you point me to where in the code base this happens (I can't find it). And then do we still want to try and download them to the host?

**Comment by kdbinder** [comment]
I would leave the logs in cloud storage (similar to other methodologies used in the past). I dont believe its necessary to download them to the local workstation.

you can simply print a message to the user to check their GCS bucket for diagnostics

**Comment by vsoch** [comment]
@kdbinder that is my intuition as well - we would want to package these logs in the context of Google Life Sciences (and not implement something alongside the general "get the output files and download to the host) to complicate the logic to finish up. I wonder if we could have some basic `--google-life-sciences-logs <src> <dest>` flag that would work with a snakemake exec call (on the instance, without the life sciences prefix) to know to package up files in `<src>` and upload them to a `<dest>` (which should be a Google storage bucket and prefix). @johanneskoester where in snakemake do we do cleaning up / finishing / upload of current logs?

**Comment by aryam7** [comment]
> the snakemake exec command is currently the last thing we run for the container entrypoint, so should there be a flag added to that to --save-google-lifesciences-logs

Just jumping in for a quick note:
If the action running the Snakemake container fails before getting to the snakemake exec command (or if the snakemake exec command fails for whatever reason), the pipeline will exit prematurely and the logs might not get uploaded correctly!
Another option is to do the copying of the logs in a separate life sciences action in which [`always_run` is set to true](https://cloud.google.com/life-sciences/docs/reference/rpc/google.cloud.lifesciences.v2beta#google.cloud.lifesciences.v2beta.Action).

**Comment by vsoch** [comment]
@aryam7 I think that would work! That does seem to include all logs that were generated for the entire pipeline (including other actions). 

![image](https://user-images.githubusercontent.com/814322/91910066-9396a280-ec6b-11ea-8123-5025ab83a5f9.png)

**Comment by aryam7** [comment]
As for whether to store the logs in the `.snakemake` folder locally or somewhere in the storage bucket, my vote is for the storage bucket.
As @kdbinder suggested, I like the idea of printing a message about where the logs are. I think it might be even more convenient if the message was the `gsutil cat /google/logs/action/<action>/output` command with the path to the logs and the action ID already printed. But that's just a minor thing!

**Comment by johanneskoester** [comment]
> As for whether to store the logs in the `.snakemake` folder locally or somewhere in the storage bucket, my vote is for the storage bucket.
> As @kdbinder suggested, I like the idea of printing a message about where the logs are. I think it might be even more convenient if the message was the `gsutil cat /google/logs/action/<action>/output` command with the path to the logs and the action ID already printed. But that's just a minor thing!

I completely agree now. That's a good idea. (also downloading to local would probably cause additional unnecessary costs).

However, let us go one step back and think about why this is necessary. My intuition was that `gcloud beta lifescience operations describe` should print the entire stdout and stderr. In combination with the log file defined in the rule itself this should already cover the entire logging output. If that is not the case, (I think @kdbinder said that `describe` output can be incomplete) we should first find out whether that is a bug or intended. @moschetti, can you comment?

**Comment by aryam7** [comment]
It might be [in the docs](https://cloud.google.com/life-sciences/docs/reference/rpc/google.cloud.lifesciences.v2beta#containerstoppedevent).
> Note that only a small amount of the end of the stream is captured here. The entire stream is stored in the /google/logs directory mounted into each action, and can be copied off the machine as described elsewhere.

By the way, my own experience has been that the [output in the log directive's file is _always_ empty](https://github.com/snakemake/snakemake/issues/589). When I first submitted this feature request, I think this was happening because the container containing Snakemake was getting stopped before Snakemake could copy its logs, but I've now discovered that this happens regardless. Since I think both situations can cause users to lose log output, I'm proposing that both should be addressed.

**Comment by moschetti** [comment]
yeah, as mentioned, the describe method not dumping the entire stdout/err is intended. they should be copied separateli.

**Comment by vsoch** [comment]
See https://github.com/snakemake/snakemake/pull/597

**Comment by johanneskoester** [comment]
> It might be [in the docs](https://cloud.google.com/life-sciences/docs/reference/rpc/google.cloud.lifesciences.v2beta#containerstoppedevent).
> 
> > Note that only a small amount of the end of the stream is captured here. The entire stream is stored in the /google/logs directory mounted into each action, and can be copied off the machine as described elsewhere.
> 
> By the way, my own experience has been that the [output in the log directive's file is _always_ empty](https://github.com/snakemake/snakemake/issues/589). When I first requested better logging, I think this was happening because the container containing Snakemake was getting stopped before Snakemake could copy its logs (because of an error occurring before Snakemake was executed within the container). But I've now discovered that this happens even for successfully completed rules, for some reason. Since I think both situations can cause users to lose log output, I'm proposing that both should be addressed. But just addressing this one would help tremendously!

@aryam7 I have answered to your issue #589. Not seeing anything in the log files is just a symptom of not using them in the right way in your example. Nevertheless, it is good to provide a way to capture the google logs as well, although they should in fact be rarely needed when using per-rule log files properly.

**Comment by johanneskoester** [comment]
Closing as @vsoch's PR is merged.

**Comment by johanneskoester** [comment]
nonsense, this is the meta-issue and should remain open :-)

**Comment by shukwong** [comment]
Hi, not sure if this is the right place to report a potential feature improvement. I noticed that I have to name the snakemake file Snakefile to work with --google-lifesciences, the option works fine with local and cluster mode. 

#this is fine
$ snakemake  --google-lifesciences --default-remote-prefix <mybucket> --google-lifesciences-region us-east1
Building DAG of jobs...

#this is fine
$ snakemake -s Snakefile --google-lifesciences --default-remote-prefix <mybucket> --google-lifesciences-region us-east1
Building DAG of jobs...

#this gives an error
$ mv Snakefile Snakefile1
$ snakemake -s Snakefile1 --google-lifesciences --default-remote-prefix <mybucket>   --google-lifesciences-region us-east1
Building DAG of jobs...
Traceback (most recent call last):
  File "/home/<myusername>/miniconda3/envs/snakemake/lib/python3.7/site-packages/snakemake/__init__.py", line 709, in snakemake
    keepincomplete=keep_incomplete,
  File "/home/<myusername>/miniconda3/envs/snakemake/lib/python3.7/site-packages/snakemake/workflow.py", line 897, in execute
    scheduler_type=scheduler_type,
  File "/home/<myusername>/miniconda3/envs/snakemake/lib/python3.7/site-packages/snakemake/scheduler.py", line 296, in __init__
    latency_wait=latency_wait,
  File "/home/<myusername>/miniconda3/envs/snakemake/lib/python3.7/site-packages/snakemake/executors/google_lifesciences.py", line 77, in __init__
    "--notemp --no-hooks --nolock " % self.snakefile
AttributeError: 'GoogleLifeSciencesExecutor' object has no attribute 'snakefile'

**Comment by vsoch** [comment]
Reporting here is good! What version of Snakemake is this, and have you tried a more recent? There indeed was a bug related to non-standard names For Snakefiles but it’s been fixed, and now this function https://github.com/snakemake/snakemake/blob/60be605c6c7c4a63abeb2c9a7e7ce95498d7a8ba/snakemake/executors/google_lifesciences.py#L68 is called before that statement, which should always set that attribute.

**Comment by shukwong** [comment]
Thanks @vsoch. My bad. I was using 5.23.0 and the error indeed went away after I updated it to 5.26.1 :)

**Comment by vsoch** [comment]
No bad at all! I’m glad that I could help, and that it was an easy fix.

**Comment by shukwong** [comment]
I have the line 
container: "docker://continuumio/miniconda3"
in my Snakemake file and it works fine if I run the job locally. But when I use it with --google-lifesciences I get the following errors: 
Does --google-lifesciences work with singularity? 

Traceback (most recent call last):
  File "/lib/python3.8/site-packages/snakemake/__init__.py", line 687, in snakemake
    success = workflow.execute(
  File "/lib/python3.8/site-packages/snakemake/workflow.py", line 1005, in execute
    success = scheduler.schedule()
  File "/lib/python3.8/site-packages/snakemake/scheduler.py", line 489, in schedule
    self.run(runjobs)
  File "/lib/python3.8/site-packages/snakemake/scheduler.py", line 500, in run
    executor.run_jobs(
  File "/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 131, in run_jobs
    self.run(
  File "/lib/python3.8/site-packages/snakemake/executors/google_lifesciences.py", line 796, in run
    if job.needs_singularity and self.workflow.use_singularity:
  File "/lib/python3.8/site-packages/snakemake/jobs.py", line 337, in needs_singularity
    return self.container_img is not None
  File "/lib/python3.8/site-packages/snakemake/jobs.py", line 346, in container_img
    return self.dag.container_imgs[self.container_img_url]
KeyError: 'docker://continuumio/miniconda3'

**Comment by vsoch** [comment]
hey @shukwong ! You are correct that the GLS executor does not support Singularity, and this is because we don't have the right permissions on the instance for it. There should have been a warning error message:

```
        # capabilities - this won't currently work (Singularity in Docker)
        # We either need to add CAPS or run in privileged mode
        if job.needs_singularity and self.workflow.use_singularity:
            logger.warning(
                "Singularity requires additional capabilities that "
                "aren't yet supported for standard Docker runs."
            )
```
But in practice likely we should exit there instead of issuing a warning, as it will lead to the error above. I'll open a PR to fix this up.

**Comment by shukwong** [comment]
Hi @vsoch, thank you so much for your super quick response :)

**Comment by vsoch** [comment]
Definitely! This should help for the future to make this more clear, thanks for reporting the issue! https://github.com/snakemake/snakemake/pull/822

**Comment by shukwong** [comment]
I have another question about the GCP Life Sciences Executor. I got my own pipeline working but it seems that it execute one job at a time (sequentially instead of in parallel) even though there are multiple independent samples.  I don't know if that's related to Snakemake or my life sciences API set up?  It does submit jobs in parallel with the cluster mode.

**Comment by vsoch** [comment]
I'm not familiar with the cluster executor so I don't have the point of comparison. Could you share your snakefile here and @johanneskoester can take a look and describe the expected outcome, given GLS?

**Comment by shukwong** [comment]
We could just use the tutorial example:
https://snakemake.readthedocs.io/en/stable/executor_tutorial/google_lifesciences.html
I would expect that the first step bwa_map is done on sample A and sample B independently and spinning up two VM instances on GCP. But it seems that it started a VM to run rule bwa_map on sample B, then samtools_sort on the output from bwa_map on sample B, before starting bwa_map on sample A and so on.  And it looks like there's at most one VM running for this pipeline.

**Comment by vsoch** [comment]
@johanneskoester shouldn't snakemake handle the concurrency of the executor?

**Comment by vsoch** [comment]
@shukwong just for a sanity check - are you setting `--jobs` ?

**Comment by shukwong** [comment]
OMG! that's it! I knew it must have to do with me :) I just set --jobs 100 and it now works beautifully 👍  Thanks again for your help @vsoch !

**Comment by vsoch** [comment]
haha, awesome! 😎  And sorry it didn’t come to me immediately, I never do anything substantially scaled when developing so I totally forgot about it :) I would bet a cluster submitter is using a job queue so it might (by default) have a higher limit.

**Comment by BobFukkink** [comment]
Hi @vsoch,

I am trying to run snakemake in the cloud with google-life-sciences and using Singularity containers. I get the same KeyError as @shukwong did, where my containers are not recognized. When I run on a local cluster, the container works fine.

So I don't use Singularity in a Docker container but just Singularity containers themselves. 

I was wondering if the Singularity with google-life-sciences compatibility is still an issue or if there is a solution for it?

Thanks in advance!

**Comment by vsoch** [comment]
Singularity containers wont work with Google Life Sciences (the last time I asked at least) because they need permissions that the VMs cannot provide. So your best bet is to use a Docker base or just the default. And remember that your base image will need to have snakemake.

**Comment by BobFukkink** [comment]
@vsoch Thank you for your quick response!

I'm sad to hear that Singularity and GLS are not compatible yet but indeed will try Docker instead.

**Comment by vsoch** [comment]
Sure thing! And I wouldn't count on ever having support. I suspect that there isn't a compelling need for Singularity from Google's perspective so I don't think it will be something worked on anytime soon (or ever).

**Comment by shukwong** [comment]
wow, I did not know that Docker would work with Google Life Sciences, that's great news!  It's just that the base image has to have snakemake or the docker image for each rule would also need to have snakemake?

**Comment by vsoch** [comment]
To be more specific, you are always able to define the `--container-image` (which defaults to the latest snakemake) and then typically use conda or (a non container) inside. It doesn't work on GLS to run a container from within a container to the best of my knowledge.

**Comment by vsoch** [comment]
The entrypoint will attempt to activate an environment called snakemake here https://github.com/snakemake/snakemake/blob/9a07cd90f149849cc815e2d342daffbe0db638fb/snakemake/executors/google_lifesciences.py#L690 and just skip if it's not found. So you can use that as a base and then use conda / wrappers in your workflow, just have snakemake available without sourcing anything, or do a custom container with the same source. I would expect most people just use the default container base.

**Comment by shukwong** [comment]
I see. Thanks @vsoch ! I have a pipeline that needs to support slurm, SGE and GCP.   I am considering switching to running via kubernetes on GCP so I can use containers on all 3 instead.

**Comment by BobFukkink** [comment]
@vsoch I am having some trouble using Docker instead of Singularity container. 

The base environment I am using is a conda environment with Snakemake, GCP, python, and some other dependencies. From that environment I run the Snakemake command **locally**:

`snakemake -s Snakefile -p --use-singularity --singularity-args "-B '/mnt' --home $HOME" --cores 1`

Within my Snakemake rule I specified the container as:
`singularity:
        docker://biocontainers/fastqc:v0.11.9_cv8`

This works well and the same as when I used a Singulariy container before. 

Now when I try to run this in the **cloud** with GCP using the following command:
`snakemake -s Snakefile -p --use-singularity --singularity-args "-B '/mnt' --home $HOME" --cores 1 --google-lifesciences --default-remote-prefix test_data --google-lifesciences-region europe-west4 --default-resources machinetype=n1-standard-1`

The same keyerror occurs as before "KeyError: 'docker://biocontainers/fastqc:v0.11.9_cv8'". 

I assume that I should not use the "--use-singularity" option in my run command but I am uncertain what to use instead?
The only info I can find about Docker containers for Snakemake is using Kubernetes. Might that be the solution?
And the "--container-image" would not work since I want to use a different container for every rule, right?

I am new to using GCP so perhaps I am doing some clumsy things. ;p

**Comment by vsoch** [comment]
> The same keyerror occurs as before "KeyError: 'docker://biocontainers/fastqc:v0.11.9_cv8'".

@BobFukkink for Google Life Sciences you can specify a base image with `--container-image` that snakemake and dependencies must be installed in, and this will be used across nodes. You cannot run containers within that - so the example that you are describing will not work. It could be possible to modify snakemake to be able to take a different container per rule, but then you would need to develop / deploy one for each step! I don't think I would want to do this or recommend it.

> I assume that I should not use the "--use-singularity" option in my run command but I am uncertain what to use instead?

I would suggest `--use-conda` - that combined with snakemake wrappers https://snakemake-wrappers.readthedocs.io/en/stable/ can make a relatively reproducible workflow.

> The only info I can find about Docker containers for Snakemake is using Kubernetes. Might that be the solution? And the "--container-image" would not work since I want to use a different container for every rule, right?

I haven't used kubernetes mysefl for Snakemake, but reading at the end of this section: https://snakemake.readthedocs.io/en/stable/snakefiles/deployment.html#running-jobs-in-containers it does look like Kubernetes would support running a Singularity container (in privileged mode) so that would be something to try.

**Comment by BobFukkink** [comment]
@vsoch The reason I want to use a different container for each rule is since I am using different tools for echt step. For instance, a rule for FastQC, Trimmomatic, BWA, GATK, etc. If I would put all of the tools in one single container together with Snakemake and other dependencies, you would get a container of ~10 GB or higher. If you would use that as a base image, each rule would use the ~10GB image and use memory and perhaps compute resources that are unessential to that step in the pipeline. Making it more expensive over time. So to my understanding, you would prefer to use a lot of images that are as small as possible to not waste your resources. Or am I wrong about this and is it possible to put it all in one?

The --use-conda would be my next option as well, but the reproducibility is a bit better with Docker or Singularity so therefore my preference, if possible of course. ;)

**Comment by vsoch** [comment]
@BobFukkink have you taken a look at the wrappers?

**Comment by BobFukkink** [comment]
@vsoch Yes, that seems to work well. I will just proceed using that instead of Singularity or Docker.

Thank you for your time and advice!

**Comment by johanneskoester** [comment]
> @vsoch The reason I want to use a different container for each rule is since I am using different tools for echt step. For instance, a rule for FastQC, Trimmomatic, BWA, GATK, etc. If I would put all of the tools in one single container together with Snakemake and other dependencies, you would get a container of ~10 GB or higher. If you would use that as a base image, each rule would use the ~10GB image and use memory and perhaps compute resources that are unessential to that step in the pipeline. Making it more expensive over time. So to my understanding, you would prefer to use a lot of images that are as small as possible to not waste your resources. Or am I wrong about this and is it possible to put it all in one?
> 
> The --use-conda would be my next option as well, but the reproducibility is a bit better with Docker or Singularity so therefore my preference, if possible of course. ;)

Yes, exactly the reason why conda is so good. It runs inside of one base container. In combination with mamba as a frontend it is really fast and robust, without all the overhead that containers generate.
